crates/rpc/rpc/src/eth/api/transactions.rs

```diff
@@ -865,6 +881,108 @@ where
         };
         build_transaction_receipt_with_block_receipts(tx, meta, receipt, &all_receipts)
     }
+
+    /// Helper function for `eth_getTransactionReceipt` (optimism)
+    ///
+    /// Returns the receipt
+    #[cfg(feature = "optimism")]
+    pub(crate) async fn build_transaction_receipt(
+        &self,
+        tx: TransactionSigned,
+        meta: TransactionMeta,
+        receipt: Receipt,
+    ) -> EthResult<TransactionReceipt> {
+        // get all receipts for the block
+        let all_receipts = match self.cache().get_receipts(meta.block_hash).await? {
+            Some(recpts) => recpts,
+            None => return Err(EthApiError::UnknownBlockNumber),
+        };
+        let block = self
+            .provider()
+            .block_by_number(meta.block_number)?
+            .ok_or(EthApiError::UnknownBlockNumber)?;
```

Comment:
this should be fetched async from the cache just like the receipt, but concurrently, via try_join

see https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/cache/mod.rs#L166-L166

for example